src/libertem/executor/pipelined.py

```diff
@@ -816,17 +942,33 @@ def yield_result_if_found(block, timeout):
                         result_task_id = result["task_id"]
                         yield (result["result"], id_to_task[result_task_id], result_task_id)
                         del id_to_task[result_task_id]
-                        assert len(id_to_task) == in_flight[0]
+                        if len(id_to_task) != in_flight[0]:
+                            raise RuntimeError(
+                                "state mismatch; `id_to_task` mapping should match `in_flight`"
+                            )
                 except WorkerQueueEmpty:
                     self._validate_worker_state()
 
             for task_idx, task in enumerate(tasks):
+                # important: call `schedule_task` before incrementing number of
```

Comment:
If this is the case then should we only increment `in_flight` and add to `id_to_task` *after* the `'RUN_TASK'` message is sent (and maybe even after `'END_TASK'` too ?), in case for some reason sending the task fails (queue closes, process crashes etc...)

Reply:
Yup, moved things around a bit